include/nbl/builtin/hlsl/scan/virtual_workgroup.hlsl

```diff
@@ -2,91 +2,85 @@
+        // This was called inside getData but due to inRange it can possibly diverge 
+        // so it must be called outside! Not yet sure why a mem barrier must be uniformly called.
+        glsl::memoryBarrierBuffer(); // scanScratchBuf can't be declared as coherent due to VMM(?)
+        
+        Storage_t data = Binop::identity; // REVIEW: replace Storage_t with Binop::type_t?
+        if(inRange)
+        {
+            getData<Storage_t, isExclusive>(data, levelInvocationIndex, levelWorkgroupIndex, treeLevel, pseudoLevel);
+        }
 
-		const bool inRange = levelInvocationIndex <= params.lastElement[pseudoLevel];
+        bool doReduce = isScan ? treeLevel < params.topLevel : treeLevel <= params.topLevel;
 
-		Storage_t data = Binop::identity();
-		if(inRange)
-		{
-			getData(data, levelInvocationIndex, localWorkgroupIndex, treeLevel, pseudoLevel);
-		}
+        if(doReduce)
+        {
+            data = workgroup::reduction<Binop,WorkgroupSize,device_capabilities>::template __call<Accessor>(data,accessor);
+        }
+        else if (!isExclusive && params.topLevel == 0u)
+        {
+            data = workgroup::inclusive_scan<Binop,WorkgroupSize,device_capabilities>::template __call<Accessor>(data,accessor);
+        }
+        else if (treeLevel != params.topLevel)
+        {
+            data = workgroup::inclusive_scan<Binop,WorkgroupSize,device_capabilities>::template __call<Accessor>(data,accessor);
+        }
+        else
+        {
+            data = workgroup::exclusive_scan<Binop,WorkgroupSize,device_capabilities>::template __call<Accessor>(data,accessor);
+        }
+        setData<Storage_t, isScan>(data, levelInvocationIndex, levelWorkgroupIndex, treeLevel, pseudoLevel, inRange);
+    }
 
-		if(treeLevel < params.topLevel) 
-		{
-			#error "Must also define some scratch accessor when calling operation()"
-			data = workgroup::reduction<Binop>()(data);
-		}
-		// REVIEW: missing _TYPE_ check and extra case here
-		else if (treeLevel != params.topLevel)
-		{
-			data = workgroup::inclusive_scan<Binop>()(data);
-		}
-		else
-		{
-			data = workgroup::exclusive_scan<Binop>()(data);
-		}
-		setData(data, levelInvocationIndex, localWorkgroupIndex, treeLevel, pseudoLevel, inRange);
-	}
-}
-}
-}
+    DefaultSchedulerParameters_t getSchedulerParameters(); // this is defined in the final shader that assembles all the SCAN operation components
+    template<class Binop, typename Storage_t, bool isScan, bool isExclusive, uint16_t WorkgroupSize, class Accessor>
+    void main(NBL_REF_ARG(Accessor) accessor)
+    {
+        const Parameters_t params = getParameters();
+        const DefaultSchedulerParameters_t schedulerParams = getSchedulerParameters();
+        Scheduler<WorkgroupSize> scheduler = Scheduler<WorkgroupSize>::create(params, schedulerParams);
+        // persistent workgroups
+        while (true)
+        {
+            if (!scheduler.getWork(accessor))
+            {
+                return;
+            }
```

Comment:
1. stop with the forward declared `get` and `set`
2. this a general pattern of running virtual workgroups, I'd move it out to a general header `virtual_workgroup_executor.hlsl` outside of `scan`

because of (2) you need to template on an opaque struct that wraps up everything you want to pass to a virtual workgroup. Basically turn the virtual workgroup into a stateful lambda.


```cpp
template<typename VirtualWorkgroup, typename Scheduler>
struct virtual_workgroup_executor
{
   void operator()(NBL_REF_ARG(Scheduler) scheduler);
};
```